### ISSUE TITLE
DynamoDB connector

### ISSUE THREAD

**Issue by nikolasburk** [prisma/prisma2#1676]
This feature requests serves as a central place to discuss development and progress for the DynamoDB connector. 

Please leave a 👍 and/or a comment in this issue if you'd like to see support for DynamoDB soon.

**Comment by amesas** [comment]
This one would be great, we use dynamodb for better "less-ops" scalability.
We connect to Posgres using Prisma, and dynamodb with custom code, we would love to use prisma as our only "connector" to data.

**Comment by fbjork** [comment]
Would be great to see DynamoDB support. Single Table design support would be amazing too.

**Comment by stewartjarod** [comment]
Single-Table, Adjacency pattern: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/bp-adjacency-graphs.html#bp-adjacency-lists

**Comment by markthethomas** [comment]
This would be huge. Dynamo has a ton going for it - easy to run/scale, very very fast, offers stronger read/writes in the cases you want. Biggest drawback has always been the learning curve/lack of easier to work tools. if prisma supported this I would use it in a heartbeat over postgres. I love PG but for so many of my projects dynamo would check every box 💯

**Comment by IamManchanda** [comment]
Awaiting DynamoDB & MongoDB support

**Comment by softmarshmallow** [comment]
Any ongoing development for this?

**Comment by changtimwu** [comment]
AWS just announced PantiSQL over dynamodb.   I guess this makes DynamoDB support easier.  
https://aws.amazon.com/about-aws/whats-new/2020/11/you-now-can-use-a-sql-compatible-query-language-to-query-insert-update-and-delete-table-data-in-amazon-dynamodb/

**Comment by softmarshmallow** [comment]
Just for only query converting will be awesome since aws-sdk just sucks.

**Comment by adamelmore** [comment]
Any movement on this? I'd consider doing the heavy lifting here if it isn't already in the works.

**Comment by softmarshmallow** [comment]
@adamelmore sign me up too. I'me ready to commit. this needs to be done. AWS sucks, all other sucks. but we do need dynamodb,..

**Comment by ahmednawaz10p** [comment]
is it alive?

**Comment by cassidoo** [comment]
Any updates on this?

**Comment by softmarshmallow** [comment]
Since mongodb support is live as preview, naively seems that dynamodb would also be possible with same no-migration-support policy.

**Comment by mikestopcontinues** [comment]
@softmarshmallow DynamoDB has a very different design than mongo. Mongo is very similar to standard SQL in that you can have custom indexes, joins, etc. Whereas Dynamo requires thinking in a very strict hashKey/sortKey model, where the alphabetic order of your sortKey impacts how you can query data. This limitation impacts quite a lot of how the prisma feature set would map to Dynamo.

I'm here because I want it too, but it's not straight forward or easy. If nothing else, what it really suggests is that Prisma should allow custom adaptors, so that the Dynamo community could converge on the right way to integrate.

**Comment by flybayer** [comment]
At a Prisma event earlier this year, Prisma executives indicated Dynamo support almost certainly wouldn't come before next year.

**Comment by AllanOricil** [comment]
> @softmarshmallow DynamoDB has a very different design than mongo. Mongo is very similar to standard SQL in that you can have custom indexes, joins, etc. Whereas Dynamo requires thinking in a very strict hashKey/sortKey model, where the alphabetic order of your sortKey impacts how you can query data. This limitation impacts quite a lot of how the prisma feature set would map to Dynamo.
> 
> I'm here because I want it too, but it's not straight forward or easy. If nothing else, what it really suggests is that Prisma should allow custom adaptors, so that the Dynamo community could converge on the right way to integrate.

And it is funny that lots of devs out there are modeling their data into dynamo as if it were exactly as mongodb. Lots of tutoriais do that. And there is even a lib like mongoose, but for dynamo.

**Comment by VincentThomas06** [comment]
Any updates? Are you guys considering adding dynamodb?

**Comment by MichalBunkowski** [comment]
Any updates?

**Comment by janpio** [comment]
No. An update would manifest itself in a comment in this issue. If that did not happen, we have nothing new to share about this feature. Feel free to comment with additional use cases or motivation about why you need the feature though, these are always useful for us.

**Comment by Thibault2ss** [comment]
We use dynamoDB and would love a prisma connector, because we love prisma ❤️

**Comment by alexmarqs** [comment]
@janpio a good motivation for this is the bad experience (DX) when dealing with dynamodb client. Additionally, you don't see any good ORM for this database type out there when comparing it with the provided API of Prisma. A lot of companies are using AWS Serverless architectures with dynamodb as its persistence layer. Technically, probably it's not straightforward / easy due dynamodb design/architecture but it would be great to have some support for this 👍 , game-changer!

**Comment by junaway** [comment]
Hi all,
For those who are waiting for Prisma to support dynamodb, maybe this will help you ? https://github.com/sensedeep/dynamodb-onetable

**Comment by climber2002** [comment]
@junaway I'm using dynamodb-onetable and it works great so far

**Comment by ptejada** [comment]
There is also https://www.npmjs.com/package/dynamoose. V3 beta supports Typescript and the AWS SDK v3

**Comment by AllanOricil** [comment]
> We use dynamoDB and would love a prisma connector, because we love prisma ❤️

AWS explains in some of their docs that we should not use dynamodb like we usually work with docs in mongodb, or even tables and relationships in sql. I don't believe Prisma would be suitable for creating this abstraction layer you want because it does exactly what amazon recommends developers not to do.

**Comment by cyriaque-ovaga** [comment]
Feature ?

**Comment by gultyaev** [comment]
> > We use dynamoDB and would love a prisma connector, because we love prisma ❤️
> 
> @Thibault2ss AWS explains in some of their docs that we should not use dynamodb like we usually work with docs in mongodb, or even tables and relationships in sql. I don't believe Prisma would be suitable for creating this abstraction layer you want because it does exactly what amazon recommends developers not to do.

Could you please share the links to the statements?

**Comment by AllanOricil** [comment]
> > > We use dynamoDB and would love a prisma connector, because we love prisma ❤️
> > 
> > @Thibault2ss AWS explains in some of their docs that we should not use dynamodb like we usually work with docs in mongodb, or even tables and relationships in sql. I don't believe Prisma would be suitable for creating this abstraction layer you want because it does exactly what amazon recommends developers not to do.
> 
> Could you please share the links to the statements?

Here you can read about it
 https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/bp-general-nosql-design.html

You have to also understand how dynamo differs from mongodb in terms of data organization and optimizations.
Adding prisma to "facilitate" creating multiple tables will surely end up creating a bad no soql database architecture. I'm pretty sure rooks will apply normalization as soon as possible and create lots of tables. Making the project extremely more expensive, and less scalable. 
There are some videos I watched from an engineer from aws explaining how to design an app using a single table and storing multiple entities. It comes down to using the primary key to index your entities (tables).

**Comment by handletastic** [comment]
I would love to see this happen. I know that there are quite a lot of blockers for this to happen. Anyone knows any latest movements?

**Comment by ahmad-ali14** [comment]
although I would like to use Prisma with ddb, it just feels weird, and I don't think that the relational nature of prisma suits dd.

**Comment by jayalfredprufrock** [comment]
I agree that managing dynamodb "relationships" doesn't feel like a good fit for Prisma and if you are designing complex single-table schemas to maximize performance for ultra specific use cases, a generic layer on top like Prisma would only get in the way. That said, I often use dynamodb in microservices as a simple distributed hash map because of its speed and low cost. Just having a strongly typed client that could fetch and update documents by partition keys and GSIs would be incredibly helpful.

This may have already been thrown out as well, but DynamoDb got support for PartiQL a while back. I have no experience with it so can't comment on what exactly that brings to the table, other than implementing a driver against this "SQL compatible language" could enable other types of data stores with low effort, but don't quote me on that.

**Comment by eranartzi** [comment]
I think a lot of people use ddb the same way they use mongo, so having prisma support ddb would be amazing

**Comment by miguevelasquez1** [comment]
is there some PR about that? 👀

**Comment by dahari87** [comment]
So many people i know would love to have this connector.

**Comment by jeremygottfried** [comment]
I worked with DynamoDB for a while and created a custom ORM to make it more usable, but it works using data copies, so query consistency is a big issue. Having a declarative API like prisma would be great, but I'm guessing it would be quite a challenge to make it usable with prisma. I personally concluded that the extra cost associated with maintaining such a database is not worth the savings if you need consistent, relational data. However, it's great for non-conventional data structures.

**Comment by emilshr** [comment]
Adding another comment to increase the issue relevance!

**Comment by gersongams** [comment]
please 😭

**Comment by jeremygottfried** [comment]
Here's an example of why mapping relational data in DynamoDB is such a challenge. To get something similar to an SQL table with relationships, you create a bunch of indices, which are really copies of the data with different keys. Write transactions become expensive if you want to query a lot of columns and relationships, and you're also paying the extra expense of storing data copies. Updating or deleting a row becomes a challenge, because you also have to update every index along with the main data. One of the selling points of DynamoDB is that it's low level and dynamic, so creating a declarative API may not be worthwhile.

```ts
      const TransactItems = [
        ...additionalTransacts,
        {
          Put: {
            Item: { ...instance, ...indices, ...globalSecondaryIndices },
            TableName,
          },
        },
        ...globalIndices.map((index) => ({
          Put: {
            Item: index,
            TableName,
          },
        })),
        ...belongsToEntries.map((index) => ({
          Put: {
            Item: index,
            TableName,
          },
        })),
      ];

```

One dev might want to store the entire row (*) with the email index, while another dev might only need the user name and PK with the index, so how do you allow that level of customizability and still have a usable API? DynamoDB works best when you design tables and indices based on the data you need, rather than trying to fit it into a pre-made API structure.

**Comment by thibauld** [comment]
+1 I'd love if we could use prisma with dynamodb!

**Comment by me2resh** [comment]
Would love to see this planned in the roadmap.

**Comment by tankcdr** [comment]
Definitely would prefer to be backed with DynamoDB such that I can integrate with the rest of my stack.

**Comment by andre2dash** [comment]
Please! 🔢

**Comment by hammadkhokhar** [comment]
+1

**Comment by brunodahobsec** [comment]
+1

**Comment by AdrianBalcan** [comment]
Totally understand that can be a different approach in the implementation and also in the use, but this would be such a great feature. I believe that once this implementation will get mature enough, Prisma will be able to scale a certain architecture of multiple API's which we can see that many tools and companies are heading towards.

**Comment by akshay-99h** [comment]
I have received a project recently. It has DynamoDB as database. I am accustomed to working with prisma all the time for most of my projects. I know some of the best people out there are working on this repo. I just want to know when an official support for DynamoDB will be released.

**Comment by dvins** [comment]
As pointed out by others DynamoDB is for all intents and purposes orthogonal to Prisma. It just does not have the same schema guarantees and especially query capabilities of say, MongoDB. In lieu of that, highly recommend Dynamoose as a great framework to use when working with DynamoDB. It adds the structure, type safety, and other modeling guard rails and makes a huge difference vs. hand crafted calls.

**Dynamoose**
https://dynamoosejs.com/getting_started/Introduction
https://github.com/dynamoose/dynamoose

**Comment by dtbayles** [comment]
i beg

**Comment by jquirossoto** [comment]
Consider the following library for your projects: https://www.dynamodbtoolbox.com/docs

**Comment by ThePlenkov** [comment]
@jquirossoto ORM options there are very poor comparing to what Prisma can deliver unfortunately

**Comment by Mingyang-Li** [comment]
After reading all your comments, I see a similar pattern:
1. Many teams use DynamoDB as their primary persistence layer
2. Many teams are using the same data modelling approach as relational databases in DynamoDB, despite AWS recommends against doing it
3. Many teams find `@aws-sdk/client-dynamodb` "just sucks"
4. Many teams aren't necessarily asking for Prisma to integrate DynamoDB as part of their offerings, they're asking for the "Prisma-style" developer experience that is unmatched by any other ORMs/SDKs. (In my case, my favourite Prisma feature is the `prisma.agent.findMany({ where: { firstName: { contains: "James" }, lastName: { equals: "bond", case: "INSENSITIVE" } } })` type of power filters

Given the observations above, I see these issues overlaps with most of what happens within the [**Azure CosmosDB**](https://azure.microsoft.com/en-us/products/cosmos-db) ecosystem as well.

As a result, I wrote a Prisma-style ORM / query builder catering to those who use Azure CosmosDB (NoSQL), you can check it out here: https://github.com/Mingyang-Li/cosmox. At this stage, you can use it to create, update, delete, find item by ID the **exact same way** as you'd do things in Prisma - but from CosmosDB. However, it has a strong emphasis for "finding items with complex filters" (aka `db.user.findMany({  })` type of queries)

Why did I build it? Because I wished `@azure/cosmos` was as easy to use as Prisma 😅

I see there're some awesome packages built by the folks working on [dynamoose](https://github.com/dynamoose/dynamoose) and [dynamodb-one-table](https://github.com/sensedeep/dynamodb-onetable), wouldn't you love it if one of these packages allows you to query dynamodb the same way you apply query filters using prisma?

If you're interested & are invested in building a prisma-style ORM for DynamoDB, please feel free to use [my repo](https://github.com/Mingyang-Li/cosmox) as a reference. Maybe it'll help you dynamically convert pure JS/TS object filters into DynamoDB-compatible [PartiQL queries](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ql-reference.html)

**Comment by spencerlepine** [comment]
I wrote a Prisma-style ORM / query builder catering to those who use AWS DynamoDB (NoSQL), you can check it out here: https://github.com/spencerlepine/dynamo-query

This library is currently a work-in-progress - if you have any feedback/contributions though, feel free to raise a PR

<details>

<summary>Sample Code Snippet</summary>

```typescript
import { createClient } from 'dynamo-query'; // v1.0.1-beta

type User = {
  id: string;
  firstName: string;
  lastName: string;
  age: number;
  createdAt: Date;
  isSuperAdmin: boolean;
};

const orm = createClient({
  region: 'us-east-1', // AWS region
  credentials: {
    accessKeyId: '<AWS_ACCESS_KEY_ID>',
    secretAccessKey: '<AWS_SECRET_ACCESS_KEY>',
  },
  models: t => ({
    user: t.createModel<User>({
      tableName: 'Users',
      partitionKey: 'userId',
      options: { fields: ['name', 'email', 'createdAt'] },
    })
  }),
});

const result = orm.user.findMany({ where: { firstName: { contains: 'Sam' } } });
```
</details>

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319
